### PR TITLE
feat: add end-to-end API test suite

### LINKED ISSUE #34: End-to-end functional testing
### Description

We should include an end-to-end test suite that can be executed before and after deployments to ensure all functionality is behaving properly

### Motivation

Recent 502 issues with POST requests (addressed in #33) was not caught until production deployment with manual testing. We should have a QA suite in place to detect these types of issues early, with some form of end-to-end test in a real environment

### Proposed solution

TBD

### Additional context

N/A

### PR DESCRIPTION
**Pull request by halheinemann**
### Issue reference

Closes #34

### Description of the change

Adds a new `@yme/e2e` service with vitest-based functional tests that make real HTTP requests against a configurable API endpoint. The suite is designed to run before and after deployments to catch issues like the 502 from #33.

**Test coverage:**
- Health/smoke checks for API reachability
- Posts: listing, pagination, schema validation, single-post retrieval
- Comments: listing, schema validation, auth requirements
- POST endpoints: auth enforcement (401), input validation (400), and explicit `< 500` assertions to catch 5xx regressions
- CORS headers and HTTP edge cases (malformed bodies, wrong content types)

**Design decisions:**
- Tests run against any environment via `E2E_BASE_URL` (defaults to `localhost:3001`)
- Write tests are gated behind `E2E_AUTH_TOKEN` and auto-skip when not set
- Excluded from `turbo run test` since e2e requires a running API server
- 60s timeouts on POST tests to accommodate Cloud Functions cold starts

**Usage:**
```bash
bun run test:e2e              # against local dev server
bun run test:e2e:production   # against production
```

### Additional context

Running against production revealed that all POST requests currently timeout (>60s), confirming the issue described in #33. All 12 GET-based tests pass against production. The POST test failures are real production bugs, not test bugs — exactly the kind of regression this suite is designed to catch.

### CODE REVIEW

**Comment by matootie** [comment]
There's a lot of failed checks here due to the localhost server not running can you look into a solution @halheinemann ?

**Comment by matootie** [comment]
great but can we add a system to run these e2e tests in CI?

I want to automate this
Ideally, deploy to staging environment, execute e2e tests there to ensure functionality works on firebase before merging

**Comment by halheinemann** [comment]
Added an e2e CI workflow in [`e2e.yml`](https://github.com/matootie/yoik.me/blob/feat/issue-34-e2e-tests/.github/workflows/e2e.yml). Here's how it works:

**On every PR push:**
1. Builds the client and bundles Cloud Functions (same as the deploy pipeline)
2. Deploys functions globally (they're shared — no preview channel equivalent)
3. Deploys Hosting to an ephemeral preview channel (`pr-<number>`, expires after 3 days)
4. Runs the e2e suite against the preview URL

**On PR close:** cleans up the preview channel automatically.

The preview channel URL gets the same Hosting rewrite rules, so `/api/**` routes to the deployed function — tests run against the real Firebase stack.

One caveat: since Cloud Functions don't have preview isolation, the function deploy here hits the live function endpoint. If the branch has function changes, those go live before the PR merges. For this repo's current scale that's fine, but worth knowing.

**Comment by matootie** [comment]
> One caveat: since Cloud Functions don't have preview isolation, the function deploy here hits the live function endpoint. If the branch has function changes, those go live before the PR merges. For this repo's current scale that's fine, but worth knowing.

This is not going to work. These functions are going to be hooked up to real production data which poses a risk to end users. Of course we can ensure only seeded test data is used and gets deleted after the test is complete. But we should just deploy the functions separately first, and remove them after?

We need db isolation (these tests cant use the same firestore tables) and it also causes room for error if we start changing production APIs on feature branches with these deploys (by mistake)

**Comment by halheinemann** [comment]
You're right — deploying functions to production on every PR push is a non-starter once real user data is involved. Two concrete problems:

1. **No function isolation** — `firebase deploy --only functions` overwrites the live function endpoint, so any branch with function changes mutates production before merge.
2. **No DB isolation** — e2e tests would read/write the same Firestore collections as real users.

**Proposed fix:** Switch the CI workflow to use the **Firebase Emulator Suite** (functions + Firestore emulators). The e2e tests run against the emulator on the CI runner — no production deploys, no shared data, completely isolated per run. Hosting preview channels can stay for visual review since those are already scoped per-PR.

I'll rework the workflow and test setup to use emulators. Will push an update to this branch.

**Comment by matootie** [comment]
looks good so far but can you rebase trunk into this branch so it's up to date before we merge?

**Comment by halheinemann** [comment]
Rebased onto trunk. Ready when you are.

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+      - name: Setup Bun
+        uses: oven-sh/setup-bun@v2
+        with:
+          bun-version: "1.3.8"
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: "20"
```
Comment (matootie) [inline .github/workflows/e2e.yml]: Didn't we face an issue with this in a previous change? Should we be using a later version of node?

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        with:
+          credentials_json: ${{ secrets.FIREBASE_SERVICE_ACCOUNT }}
+
+      - name: Deploy Hosting to preview channel
+        if: ${{ !cancelled() }}
+        run: |
+          bunx firebase-tools hosting:channel:deploy \
+            pr-${{ github.event.pull_request.number }} \
+            --expires 3d \
+            --project yoikme
```
Comment (matootie) [inline .github/workflows/e2e.yml]: So now that we're running e2e tests using an emulator this preview deployment step has nothing to do with test execution and can be removed from this workflow

I would suggest creating a separate workflow for preview deployments whenever there's front end changes in a PR

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            --project yoikme &
+          # Wait for the hosting emulator to accept connections.
+          timeout 120 bash -c \
+            'until curl -sf http://127.0.0.1:5000 > /dev/null 2>&1; do sleep 2; done'
+          echo "Emulators ready"
+
+      - name: Run E2E tests
+        run: cd services/e2e && bunx vitest run
+        env:
+          E2E_BASE_URL: http://127.0.0.1:5000
```
Comment (matootie) [inline .github/workflows/e2e.yml:81]: Should set auth token here since we're using the emulator so some of our test cases don't get skipped, right?

**Review by matootie — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e.yml (hunk 1 above)
Comment (halheinemann) [reply .github/workflows/e2e.yml]: Good catch — updated to Node 22 to match the `nodejs22` runtime in `firebase.json` and the engine requirement in `functions/package.json`.

File: .github/workflows/e2e.yml (hunk 2 above)
Comment (halheinemann) [reply .github/workflows/e2e.yml]: Agreed. Moved the preview deployment into a dedicated `.github/workflows/preview.yml` that only triggers on `services/client/**` changes. The cleanup job is included there as well.

File: .github/workflows/e2e.yml (hunk 3 above)
Comment (halheinemann) [reply .github/workflows/e2e.yml:81]: Done — added the auth emulator to `firebase.json` and the `--only` flag, then added a step that creates a test user via the Auth emulator REST API and passes the `idToken` as `E2E_AUTH_TOKEN` to the test step. Also set `FIREBASE_AUTH_EMULATOR_HOST` in the test env for good measure. No more skipped auth-gated tests in CI.
